### DISCUSSION TITLE
delay(1) needed in while() loop for ESP32-S3

### DISCUSSION THREAD

**Discussion by aviateur17** [timmbogner/Farm-Data-Relay-System#169] [Q&A]
So I've been banging my head on the wall for awhile and can't figure out what is going on so I'm hoping someone can help.  For the code listed below if I comment out the delay(1) line the output seems to hang at that point but does not go into watchdog timeout.  It will just not proceed past that line - the serial shell output just outputs the previous DBG and nothing else.  I've tried to search for the answer but cannot find a thing.   So, for ESP32 why is that delay(1) line necessary?   delay(0) and yield() do not seem to work - delay(1) is required;  but why???   Note that when testing this the gateway that it is trying to ping is not alive.  But what I expect it to do is time out at the given 300ms but that does not happen. It seems to just hang.  Also, if I replace the delay(1) with a DBG statement it also seems to work without issue.   I just can't understand why that stuff is needed.

MCU is a Seeed Xiao ESP32-S3

```
uint32_t pingFDRSEspNow(uint8_t *address, uint32_t timeout) {
    SystemPacket sys_packet = {.cmd = cmd_ping, .param = 0};
    
    pingFlag = false;
    DBG(" ESP-NOW ping sent.");
    esp_now_send(address, (uint8_t *)&sys_packet, sizeof(SystemPacket));
    time_t ping_start = millis();
    while ((millis() - ping_start) <= timeout)
    {
        delay(1); // For ESP32 if this delay isn't present it will hang here??
        // yield(); // do I need to yield or does it automatically?
        if (pingFlag)
        {
            DBG("ESP-NOW Ping Reply in " + String(millis() - ping_start) + "ms from 0x" + String(address[5], HEX));
            return (millis() - ping_start);
        }
    }
    DBG("No ESP-NOW ping returned within " + String(timeout) + "ms.");
    return UINT32_MAX;
}
```

**Comment by aviateur17** [comment]
Did some testing and here are my results.  Mostly for my future recollection as my head hurts this morning.

ESP32-S3:  delay(1) seems to be required in while loops
ESP32:  Neither delay() nor yield() required in any while loops
ESP8266:  delay(1) OR yield() required in while loops

not sure if anyone wants to do testing on the ESP32-S2 or ESP32-C3 but I would guess that they are like the ESP32-S3.

Hence:

I would propose that we use delay(1) in while loops for compatibility purposes.  I don't think we should do precompiler preprocessing based on MCU because delay(1) won't have any appreciable effect on the program running in other microcontrollers.

Thoughts?

**Reply by timmbogner** [reply]
Well take it easy on your noggin when I tell you this: I can't seem to recreate this problem. I'm using a Heltec LoRa32v3, which has the S3 chip, and when I run the ESPNOW_Controller it either fails the ping and moves on to failing to register, or it pings successfully and then registers okay. 

Coincidentally, I have several ESP32-S3 devices that shipped today from LilyGo, so I'll have a bit more to work with in a week or so. No Xiaos, though.

**Reply by timmbogner** [reply]
Note: I dropped in your exact code to replace my current version of the function, but I still can't get it to hang.

**Reply by timmbogner** [reply]
On the subject of new devices: LilyGo is sending me samples of their new LoRa board, T3-S3, in both SX1262 and SX1280.
In the spirit of this, I was thinking it would be cool to send you (Jeff) a package with some surplus gear of mine, including an old LoRa32v1 and various modules that I inexplicably bought way more than I needed of.
If you're interested, send me your mailing info via my email.

**Reply by aviateur17** [reply]
> Note: I dropped in your exact code to replace my current version of the function, but I still can't get it to hang.

Timm, thanks for testing.  Just to make sure, you used my code and removed the delay(1) or commented it out and had no issues?   I guess I haven't looked to see how it is in the main branch currently but I'm guessing there is no delay(1).   I'm not sure if there is a difference in the compiler or what would be going on between my device requiring the delay and yours not.

**Reply by timmbogner** [reply]
Yeah, I tried both with and without the delay and I couldn't get it to hang. There's no delay in the main branch, but there is the yield() for 8266.
I'm suspicious that it maybe be specific to the Xiao somehow? I can do more troubleshooting tonight with my S3.

**Reply by aviateur17** [reply]
Okay, thanks for confirming.  I wouldn't waste any more time on it.  I'll just put it in the code on my side until I can figure out what's causing it on my device.  I'll send you that email shortly on the devices.   I just received those small solar panels and mppt boards from AliExpress so I have to try those out along with hooking up the RTC module and GPS.  So lots of new stuff to try just not enough time to do it all.

**Comment by bvwelch** [comment]
Greetings,
I would like to learn how to use JTAG with a Xiao esp32-c3 anyhow, so maybe I can test this issue at the same time. Is there an existing example program that can be used as-is, or with minor edit? maybe using platformio? thanks.

**Reply by timmbogner** [reply]
@aviateur17 can jump in with any corrections, but:

- If the default ESPNOW_Controller example works on your Xiao, then there won't be anything to investigate.
- If the example hangs when sending a ping during setup(), then that's the problem we're looking into.

**Comment by aviateur17** [comment]
I'd agree with Timm, seems like this issue is really only on my end.  If you want to spend time that would be awesome but it's not necessary.  It would be a great way to test out JTAG functionality.

I've meant to try this but the JTAG hardware requires 6 additional pins and the hardware.  There are different kinds of hardware you can use but the most cost efficient is ESP-PROG which you can find on Ali express and probably Amazon or other places.   I'll link to a video of a guy I really like demonstrating the use of JTAG and PlatformIO.   The Xiao ESP32-S3 has 6 solder pads on the back that can be used but they are pretty tiny.   I'm not sure if the C3 has the same.  Yeah the C3 has the same pads.  You'd need to solder wire or make some kind of header to hook up the cable from the ESP-PROG:
![image](https://github.com/timmbogner/Farm-Data-Relay-System/assets/61305397/83c041c7-ce2d-456e-8dad-f89da6b72a9c)


PlatformIO JTAG debugging:
https://docs.platformio.org/en/latest/platforms/espressif32.html#debugging

Video demonstrating how to set up and perform debugging:
https://youtu.be/TivyIFF-dzw

**Comment by bvwelch** [comment]
Thanks i look forward to seeing platformio video. I have used jtag many
times but never on esp32, and apparently it is built-in to the C3 usb cable.

**Reply by aviateur17** [reply]
That would be sweet to not have to use the ESP-PROG hardware.  I haven't looked very hard but if you could send a link on how to use the built in USB interface that would be awesome.  Thanks!

**Comment by bvwelch** [comment]
Well good news and bad news -- the good news is that it looks like the built-in usb cable works for JTAG, but I read a warning that during deep sleep the USB will be turned off, and I guess that would mean JTAG will fail.

So, I guess I am even more interested in your videos! :-)

Note I haven't actually tested jtag with the usb cable yet, here is what I have so far:

(ubuntu system)

hab$ pio pkg exec -p tool-openocd-esp32 -c "openocd -f board/esp32c3-builtin.cfg"
Using tool-openocd-esp32@2.1200.20230419 package
Open On-Chip Debugger v0.12.0-esp32-20230419 (2023-04-18-22:02)
Licensed under GNU GPL v2
For bug reports, read
	http://openocd.org/doc/doxygen/bugs.html
Info : only one transport option; autoselecting 'jtag'
Info : esp_usb_jtag: VID set to 0x303a and PID to 0x1001
Info : esp_usb_jtag: capabilities descriptor set to 0x2000
Warn : Transport "jtag" was already selected
Info : Listening on port 6666 for tcl connections
Info : Listening on port 4444 for telnet connections
Info : esp_usb_jtag: serial (34:85:18:05:36:B0)
Info : esp_usb_jtag: Device found. Base speed 40000KHz, div range 1 to 255
Info : clock speed 40000 kHz
Info : JTAG tap: esp32c3.cpu tap/device found: 0x00005c25 (mfg: 0x612 (Espressif Systems), part: 0x0005, ver: 0x0)
Info : datacount=2 progbufsize=16
Info : Examined RISC-V core; found 1 harts
Info :  hart 0: XLEN=32, misa=0x40101104
Info : starting gdb server for esp32c3 on 3333
Info : Listening on port 3333 for gdb connections

-bill

**Comment by bvwelch** [comment]
And by the way, not using jtag, just the arduino ide and the example sketch, here is what I see on the serial monitor:


FDRS User Node initializing...
 Reading ID 1
 Gateway: 1
Initializing ESP-NOW!
 ESP-NOW Initialized.
 ESP-NOW ping sent.
No ESP-NOW ping returned within 1000ms.
ESP-NOW peer registration request submitted to 1
No gateways accepted the request
Subs

**Comment by aviateur17** [comment]
Thanks for testing it out!  Yours works too just like Timm without issue.  Must be something on my end and I'll try to figure it out at some point.  Thanks!

**Comment by bvwelch** [comment]
By the way, a reference to the built-in jtag in the C3:

https://www.espressif.com/sites/default/files/documentation/esp32-c3_technical_reference_manual_en.pdf#usbserialjtag

And note-- besides just using the same usb cable that is used for arduino ide, etc, there are two pads for a dedicated jtag usb. I havent tried that yet.

But I guess at this point I should cobble an external jtag interface that I can use with other esp32 modules besides the C3.

**Reply by bvwelch** [reply]
The esp-prog looks like a handy choice, but I will try some stuff from my junk box first -- I have some boards I have used with openocd on other projects.
